subsys/bluetooth/mesh/sensor_types.c

```diff
@@ -710,17 +996,102 @@ static bool exp_1_1_delta_check(const struct bt_mesh_sensor_value *current,
 	return exp_1_1_ceil_abs_diff(prev_raw, curr_raw) > *(delta->raw);
 }
 
-static float float32_to_float(const struct bt_mesh_sensor_value *sensor_val)
+static enum bt_mesh_sensor_value_status
+exp_1_1_to_float(const struct bt_mesh_sensor_value *sensor_val, float *val)
 {
-	uint32_t raw = sys_get_le32(sensor_val->raw);
+	uint8_t raw = *(sensor_val->raw);
 
-	return *(float *)(&raw);
+	if (raw == 0xff) {
+		return BT_MESH_SENSOR_VALUE_UNKNOWN;
+	}
+
+	if (raw == 0xfe) {
+		return BT_MESH_SENSOR_VALUE_TOTAL_DEVICE_LIFE;
+	}
+
+	if (val) {
+		*val = exp_1_1_decode_float(raw);
+	}
+	return BT_MESH_SENSOR_VALUE_NUMBER;
+}
+
+static int exp_1_1_from_float(const struct bt_mesh_sensor_format *format,
+			      float val, struct bt_mesh_sensor_value *sensor_val)
+{
+	sensor_val->format = format;
+	if (val == 0.0f) {
+		*(sensor_val->raw) = 0x00;
+		return 0;
+	}
+
+	float result = roundf(logf(val)/LOG_1_1);
+
+	*(sensor_val->raw) = CLAMP(result, 0x01, 0xfd);
+
+	if (result < 1 || result > 253) {
+		return -ERANGE;
+	}
+
+	return 0;
+}
+
+static int exp_1_1_from_special(const struct bt_mesh_sensor_format *format,
+				enum bt_mesh_sensor_value_status special,
+				struct bt_mesh_sensor_value *sensor_val)
+{
+	switch (special) {
+	case BT_MESH_SENSOR_VALUE_UNKNOWN:
+		*(sensor_val->raw) = 0xff;
+		break;
+	case BT_MESH_SENSOR_VALUE_TOTAL_DEVICE_LIFE:
+		*(sensor_val->raw) = 0xfe;
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	return 0;
+}
+
+static struct bt_mesh_sensor_format_cb exp_1_1_cb = {
+	.compare = exp_1_1_compare,
+	.delta_check = exp_1_1_delta_check,
+	.to_float = exp_1_1_to_float,
+	.from_float = exp_1_1_from_float,
+	.from_special = exp_1_1_from_special,
+};
+
+static enum bt_mesh_sensor_value_status
+float32_to_float(const struct bt_mesh_sensor_value *sensor_val, float *val)
+{
+	if (val) {
```

Comment:
Do we allow it to be NULL?

Reply:
This is the internal API. Which allows NULL, because then the public API function for just getting the status but not the value can pass NULL (the code for getting the status would otherwise be identical to this, so did it to simplify). I can clarify with a comment. The public API does not specifically allow NULL.